lib/pbench/server/unpack_tarballs.py

```diff
@@ -0,0 +1,135 @@
+import glob
+from logging import Logger
+import os
+from pathlib import Path
+import tempfile
+from typing import NamedTuple
+
+from pbench.server import PbenchServerConfig
+from pbench.server.filetree import FileTree
+from pbench.server.report import Report
+from pbench.server.utils import get_tarball_md5
+
+
+class Results(NamedTuple):
+    total: int
+    success: int
+
+
+class Unpack:
+    LINKSRC = "TO-UNPACK"
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        """Unpacks Tarball in the INCOMING Directory
+
+        Args:
+            config: PbenchServerConfig configuration object
+            logger: A Pbench python Logger
+        """
+        self.config = config
+        self.logger = logger
+
+    def unpack(self, lb_args: int, ub_args: int) -> Results:
+        """Scans for tarballs in the TO-UNPACK subdirectories of the
+        ARCHIVE directory and unpacks them using FileTree.unpack() function.
+
+        Args:
+            lb_args: Lowerbound size of Bucket
+            ub_args: Upperbound size of Bucket
+
+        Returns:
+            Results tuple containing the counts of Total and Successful tarballs.
+        """
+        tarlist = glob.iglob(self.config.ARCHIVE / "*" / self.LINKSRC / "*.tar.xz")
+
+        tarball_list = list()
+        for tarball in tarlist:
+            tar_size = Path(tarball).stat().st_size
+            if tar_size < lb_args or tar_size > ub_args:
+                continue
+            tarball_list.append(tarball)
+
+        ntotal = nsuccess = 0
+
+        file_tree = FileTree(self.config, self.logger)
+
+        for tarball in sorted(tarball_list):
+            ntotal += 1
+
+            try:
+                tb = Path(tarball).resolve(strict=True)
+            except FileNotFoundError as exc:
+                self.logger.error(
+                    "{}: Tarball link, '{}', does not resolve to a real location: {}",
+                    self.config.TS,
+                    tarball,
+                    exc,
+                )
+                continue
+
+            try:
+                tar_md5 = get_tarball_md5(tb)
+            except FileNotFoundError as exc:
+                self.logger.error(
+                    "{}: Getting md5 value of tarball '{}' failed: {}",
+                    self.config.TS,
+                    tb,
+                    exc,
+                )
+                continue
+
+            try:
+                file_tree.unpack(tar_md5)
```

Comment:
Is this unpacking a single tar ball? Shouldn't the tar ball be passed as an argument? Oherwise, how does it know which one?  I need to read the `filetree` code next :-)

Reply:
NM - I read the filetree code :-)